### PR TITLE
Add repository issue search

### PR DESCRIPTION
**Pull request by avsm**
This PR also includes a separate commit to improve error reporting (89f5eb63f3e973a54240c2260af939f3560c0208)

### CODE REVIEW

File: lib/github_core.ml [hunk 1]
```diff
@@ -1381,6 +1390,50 @@ module Make(Env : Github_s.Env)(Time : Github_s.Time)(CL : Cohttp_lwt.Client)
...
+    (*| `In of repo_issue_field list TODO *)
+      | `Author of string
+      | `Assignee of string
+      | `Mentions of string
+      | `Commenter of string
+      | `Involves of string
+      | `Team of string
+    (*| `State of repo_issue_status TODO open or closed? *)
+      | `Label of string
+      | `No_label of string
```
Comment (dsheets) [inline lib/github_core.ml]: Maybe this should be called `Without_label`?

File: lib/github_core.ml [hunk 2]
```diff
@@ -2106,6 +2159,24 @@ module Make(Env : Github_s.Env)(Time : Github_s.Time)(CL : Cohttp_lwt.Client)
...
+      let direction = Filter.string_of_direction direction in
+      let uri = URI.repo_search_issues in
+      let params = [
+        "q", q;
+        "order",direction;
+        "per_page",string_of_int 100;
+      ]@(match sort with None -> [] | Some s -> ["sort",s]) in
+      API.get_stream ~rate:Search ?token ~params ~uri (fun b ->
+        return [repository_issue_search_of_string b]
+      )
```
Comment (dsheets) [inline lib/github_core.ml]: This looks very similar to `Search.repos`. Can the few different functions be abstracted out?

File: lib/github_s.mli [hunk 3]
```diff
@@ -1635,6 +1655,14 @@ module type Github = sig
...
         [?sort] (default best match) and ordered by [?direction]
         (default [`Desc]). *)
+
+     val issues :
+       ?token:Token.t ->
+       ?sort:Filter.repo_sort ->
+       ?direction:Filter.direction ->
+       qualifiers:Filter.issue_qualifier list ->
+       keywords:string list ->
+       unit -> Github_t.repository_issue_search Stream.t
```
Comment (dsheets) [inline lib/github_s.mli]: Please add a documentation comment here.

**Comment by avsm** [comment]
I will rebase this and address the review comments later...

**Comment by samoht** [comment]
I've force pushed to @avsm's branch with a rebase + a fix for all the points listed by @dsheets

File: lib/github.atd [hunk 4]
```diff
@@ -590,7 +600,7 @@ type repo_issue_event = {
   ?label: base_label option;
   ?assignee: linked_user option;
   ?assigner: linked_user option;
-  ?milestone: milestone option;
+  ?milestone: repo_issue_milestone option;
```
Comment (dsheets) [inline lib/github.atd]: Is this change part of the feature? What is the purpose of it? Are repo issue events with milestones really using a totally different milestone type?

Comment (samoht) [inline lib/github.atd]: /cc @avsm. I have just rebased his patches and made some code cleanup :p

Comment (avsm) [inline lib/github.atd]: Yes it was wrong before. You just get a `title` back from this particular part of the API. See https://developer.github.com/v3/issues/events/#list-events-for-a-repository

> milestone
> The Milestone object including a `title` attribute. Only provided for `milestoned` and `demilestoned` events.

Roll on GraphQL...

Comment (dsheets) [inline lib/github.atd]: Great... the examples on that page have full milestone objects but I confirm that, e.g. <https://api.github.com/repos/moby/moby/issues/34222/events>, has `milestone` with only a `title`.

This change breaks type compatibility. Could you please put it in a commit by itself? It doesn't seem to be related to adding repository search.

Comment (samoht) [inline lib/github.atd]: I've force pushed a split of the last commit into two. The `milestone` field is also part of the repository events, should this be updated as well?

Comment (dsheets) [inline lib/github.atd]:
```
curl -H "Accept: application/vnd.github.mockingbird-preview" https://api.github.com/repos/moby/moby/issues/34222/timeline
```

indicates that the title-only milestone object is also returned for timeline requests. I'm not sure where you're seeing the `milestone` type used as part of repository events?

Perhaps the type should be called `milestone_reference` or similar? For objects with subsets, I also try to keep that relation fixed through the use of ATD's record inheritance feature. Maybe `milestone` should inherit from `milestone_reference`?

**Comment by samoht** [comment]
I force pushed a new version with inheritance + fix in the timeline issue. And the milestone's repository field seems to have the right type, so all is good :-)

File: lib/github.atd [hunk 5]
```diff
@@ -234,17 +234,21 @@ type update_milestone = {
   ?due_on: string option;
 } <ocaml field_prefix="update_milestone_">
 
+type milestone_reference = {
+  ~title: string;
+} <ocaml field_prefix="milestone_reference">
```
Comment (dsheets) [inline lib/github.atd]: needs `_` at end of prefix :-/

File: lib/github.atd [hunk 6]
```diff
@@ -581,6 +587,10 @@ type issue_rename = {
   to: string;
 } <ocaml field_prefix="issue_rename_">
 
+type repo_issue_milestone = {
+  title: string;
+} <ocaml field_prefix="repo_issue_milestone_">
+
```
Comment (dsheets) [inline lib/github.atd]: This shouldn't be here so shouldn't need to be removed later.

**Comment by dsheets** [comment]
Thanks! Just 2 small issues. Sorry for all the back-and-forth.

File: lib/github.atd [hunk 7]
```diff
@@ -587,10 +591,6 @@ type issue_rename = {
   to: string;
 } <ocaml field_prefix="issue_rename_">
 
-type repo_issue_milestone = {
-  title: string;
-} <ocaml field_prefix="repo_issue_milestone_">
-
```
Comment (dsheets) [inline lib/github.atd]: This shouldn't have been added in the first place?

**Comment by samoht** [comment]
Ok I think this is good now :-) Thanks for the reviews!

**Comment by samoht** [comment]
It seems that all comments of @dsheets have been addressed so merging.
